File: js/scrabble-score.js
function scoreLetter(letter){
  switch(letter) {
    case "Q":
    case "Z":
      return 10;
      break;
    case "J":
    case "X":
      return 8;
      break;
    case "K":
      return 5;
      break;
    case "F":
    case "H":
    case "V":
    case "W":
    case "Y":
      return 4;
      break;
    case "B":
    case "C":
    case "M":
    case "P":
      return 3;
      break;
    case "D":
    case "G":
      return 2;
      break;
    default:
      return 1;
  }
};

function sanitizeAndSplit(word){
  return word.toUpperCase().split("")
}

function scoreWord(letters){
  total=0
  letters.forEach(function(letter){
    total+=scoreLetter(letter);
  })
  return total
};

function scoreMultiplier(score, multiplier){
  return multiplier==="double-word" ? score * 2 : multiplier === "triple-word" ? score * 3 : score
};

module.exports = function(word, multiplier=null){
  if (word===null||word.trim()===""){return 0};
  var letters = sanitizeAndSplit(word);
  score =  scoreWord(letters);
  return scoreMultiplier(score, multiplier)
};

File: ruby/scrabble_score.rb
class Scrabble
  attr_reader :word, :multiplier

  def initialize(word, multiplier="")
    @word = word.nil? ? "" : word.strip.upcase
    @multiplier = (multiplier == :triple ? 3 : multiplier == :double ? 2 : 1)
  end

  def letter_scores
    Hash.new(1).merge({"D"=>2, "G"=>2, "B"=>3, "C"=>3, "M"=>3, "P"=>3, "F"=>4,
      "H"=>4, "V"=>4, "W"=>4, "Y"=>4, "K"=>5, "J"=>8, "X"=>8, "Q"=>10, "Z"=>10})
  end

  def points
    word.split("").map{|letter| letter_scores[letter]}.reduce(:+)
  end

  def score
    word.empty? ? 0 : (points * multiplier)
  end

  def self.score(word, multiplier="")
    Scrabble.new(word, multiplier="").score
  end
end
